### PR TITLE
1 create search api prototype

### PR DESCRIPTION
**Pull request by InnaVays**
redis index as collection + tests for parsing

### CODE REVIEW

File: app/api/utils.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+# Function to encode blocks using Cohere
+def encode_blocks(cohere_conn,text, cohere_model = COHERE_MODEL):
+    """
+    Function to encode blocks using Cohere
+    """
+    return np.array(cohere_conn.embed(texts=[text],  model=cohere_model).embeddings[0]).astype(np.float32)
+
+
+def search_by_path(redis_connect, collection_name, tag):
```
Comment (iliatimofeev) [inline app/api/utils.py:88]: In my opinion, search_by_path should be a part Collection class.

File: app/api/utils.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+VECTOR_FIELD_NAME = 'encoded_text_block'
+COHERE_MODEL = 'embed-multilingual-v2.0'
+VECTOR_DIMENSIONS = 768
+THRESHOLD = int(512/3 * 6 )
+
+def get_a_prefix(collection_name):
+    return f"{collection_name}"
+
+def get_a_key(prefix, document_key, block_num):
```
Comment (iliatimofeev) [inline app/api/utils.py:19]: In my opinion, get_a_key should be a part Collection class.

File: app/api/test_documents.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import numpy
+
+test_doc = {
```
Comment (iliatimofeev) [inline app/api/test_documents.py:3]: Please move test documents to a separate **json** file.
From #15

**Review by iliatimofeev — CHANGES_REQUESTED** [review]
** DD**
1. We can add to our collection 10 test documents including the Asq Quran.
2. When we search with query “Вопросы о боге” reference to Ask Quran returned in top 3 documents.
3. That criteria is implemented as unit test and passes
4. Search the exact part in a large document implemented as tests.

**I can't find relevant tests**

File: app/api/Depends.py
[the remark this answers is not in the document]
Comment (InnaVays) [reply app/api/Depends.py:1]: That is not my file! I did not create it, and have no idea why it exists))

File: app/api/Depends.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        return Redis( host = settings.REDIS_HOST,
+                    port = settings.REDIS_PORT,
+                    password = settings.REDIS_PASSWORD)
+
+class Collection:
+    def __init__(self, connection, name):
+        self.connection = connection
+        self.name = name
+      
+class Embedding:
```
[the remark this answers is not in the document]
Comment (InnaVays) [reply app/api/Depends.py:21]: Ok

File: app/api/api_v1/endpoints/document.py [hunk 5]
```diff
@@ -1,85 +1,110 @@
...
-    """
-    Return collection by name
-    """
-    collection = existing_collections.get(name)
-    if collection:
-        return collection
-    raise HTTPException(status_code=404, detail="Collection not found")
 
 @router.get("/{collection_name}/", response_model=List[str])
 def list_documents(collection_name: str) -> Any:
```
[the remark this answers is not in the document]
Comment (iliatimofeev) [reply app/api/api_v1/endpoints/document.py:17]: Please read this documentation finally
https://fastapi.tiangolo.com/tutorial/dependencies/

File: app/api/Depends.py
[the remark this answers is not in the document]
Comment (iliatimofeev) [reply app/api/Depends.py:1]: But it's your code in it. Just rename it.

File: app/api/api_v1/endpoints/search.py [hunk 6]
```diff
@@ -1,39 +1,81 @@
...
-    return schemas.SearchResultFull(
-        results=results,
-        suggestions=suggestions,
-        totalHits=len(results) + len(suggestions)
-    )
+    for question in questions:
+        query_embedding = utils.encode_blocks(deps.db.embedding, question)
+        # A list to store the search results for the current question from all collections
+        for collection in existing_collections:
+            #print(f"Search in {collection} ")
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/search.py:36]: Consider moving vector search into Collection class.

 ```python 
 class Collection:
       ....
       def search (self, query:str) ->List[DocumentBlocksSearchResult]
           ...
 ```
       
That will allow us to use search in other places and hide all redis/cohere specifics in one place.

**Review by iliatimofeev — COMMENTED** [review]
Redis provides a high-level API to manipulate with [Pydantic](https://pypi.org/project/pydantic/) (which are native to FastAPI) objects, take a look into [blog post](https://redis.com/blog/introducing-redis-om-for-python/) and [documentation](https://redis.io/docs/stack/get-started/tutorials/stack-python/).  Additionally there are some external packages that do it like [pydantic-redis](https://sopherapps.github.io/pydantic-redis/explanation/why-use-orms/). 

- [ ] Please explore that opportunity and consider removing low level querying and encoding in your code.

**Pull request closed without merging**